File: daily/30 -day finish 300 coding challenge/2020-5-14. 10 questions - Tree/101. Symmetric Tree.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root):
        if not root: return True
        def comparison(left, right):
            if not left and not right: return True
            if not left: return False
            if not right: return False
            if left.val != right.val: return False
            return comparison(left.left, right.right) and comparision(left.right, right.left)
        
        return comparision(root.left, root.right)

# time complexity o(N)
# space complexity o(N)

File: daily/30 -day finish 300 coding challenge/2020-5-14. 10 questions - Tree/101. Symmetric Tree.cpp
struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
};


class Solution{
public:
    bool isSymmetric(TreeNode* root){
        if(!root) return true;
        return isSymmetric(root->left, root->right);
    }
    bool isSymmetric(TreeNode* t1, TreeNode* t2){
        if(!t1 && !t2) return true;
        if(!t1 || !t2) return false;
        return t1->val == t2->val
            && isSymmetric(t1->left, t2->right) && isSymmetric(t1->right, t2->left);
    }
};